### DISCUSSION TITLE
Support user input, message history, structured output, and tool calls in prompt experiments

### DISCUSSION THREAD

**Discussion by marcklingen** [langfuse/langfuse#4454] [Ideas]
### Describe the feature or potential improvement

Currently, all variables need to be inserted via prompt variables. However, there are some standard variables that I might now want to track in Langfuse prompt management explicitly, but rather insert implicitly when starting a prompt experiment.

- input (final user message)
- message history, see related thread: https://github.com/orgs/langfuse/discussions/2210
- tool calls
- structured output

**Comment by simonwh** [comment]
+1, much needed especially when working with agents. When adding a production/live trace to a dataset, you want to include all the chat history that led to this output, and preferably without having to modify the dataset entry before saving.

**Comment by marcklingen** [comment]
Additional feedback that I received on this

> I was hoping to test different prompt versions directly within the Langfuse prompt management UI. Our prompts incorporate advanced features such as JSON schema, structured output, or function calls. While the Langfuse documentation suggests placing these details into the config, I noticed it’s not currently possible to set up experiments and/or dataset to use this config.
> 
> It would be fantastic if this capability could be added, as it would greatly enhance the flexibility and usability of the platform.

`config` currently serves as an escape hatch for various prompt-related features, but its structure lacks opinions (it is a plain json object). This highlights the need for more native support for tools and structured outputs in Langfuse prompts, which can then be relied upon in prompt experiments.

**Comment by FirminEx** [comment]
+1 Our application is all about long conversations, and the LLM tends to stop following the system prompt after a while. It would be a game changer for us to be able to test different prompts on large datasets!

**Comment by paulpacaud** [comment]
Great feature

**Comment by backpropagate404** [comment]
This is also a concern for my project where we deal with very large conversations

**Comment by simonwh** [comment]
@marcklingen Any updates on this one? 🙏

**Reply by marcklingen** [reply]
We are currently looking into it after recently releasing support for tools, functions, and structured output for the playground.

**Reply by aaparmeggiani** [reply]
@marcklingen Working great on playground.  What about experiments?

**Reply by marliessophie** [reply]
Not yet supported unfortunately and I cannot promise an exact timeline at the moment. We will keep you updated here.

**Reply by hweeminloh** [reply]
Hi there, just wanted to follow-up on this if there's a timeline to add this feature? I'm able to achieve what I'm looking to do by checking my prompt changes manually one by one using playground, against my PROD traces. Our sessions are long conversations with multiturn and it is important that I run them against the chat history. If there's a bulk way to run the same, that would be a huge time saver.

**Comment by cstalhem** [comment]
+1 on this. Especially Structured Outputs since we rely on that in our application and not having to run experiments using code would significantly speed up our process and also reduce the reliance on our main developers.
